File: data_structures/stack.py
"""
Stack (Last-In, First-Out  – LIFO)
===================================
A stack lets you add and remove items only from one end (the "top").
Think of a stack of plates: you always put a new plate on top and take
the top plate off first.

Real-world uses:
  • Undo / redo in text editors
  • Browser back / forward navigation
  • Function call stack in programming languages
  • Balanced parentheses checking

Complexities (all O(1)):
  push  – add to top
  pop   – remove from top
  peek  – look at top without removing
  is_empty – check whether the stack is empty
"""


class Stack:
    """Stack implemented with a Python list (the end of the list is the top)."""

    def __init__(self):
        self._data = []

    def push(self, item):
        """Add item to the top of the stack – O(1) amortised."""
        self._data.append(item)

    def pop(self):
        """Remove and return the top item – O(1).
        Raises IndexError if the stack is empty.
        """
        if self.is_empty():
            raise IndexError("pop from an empty stack")
        return self._data.pop()

    def peek(self):
        """Return the top item without removing it – O(1).
        Raises IndexError if the stack is empty.
        """
        if self.is_empty():
            raise IndexError("peek at an empty stack")
        return self._data[-1]

    def is_empty(self):
        """Return True if the stack has no elements – O(1)."""
        return len(self._data) == 0

    def size(self):
        """Return the number of items in the stack – O(1)."""
        return len(self._data)

    def __str__(self):
        return "Stack (top → bottom): " + str(list(reversed(self._data)))


# ── Practical Example: balanced parentheses checker ──────────────────────────

def is_balanced(expression):
    """Return True if every opening bracket has a matching closing bracket."""
    stack = Stack()
    matching = {')': '(', ']': '[', '}': '{'}

    for char in expression:
        if char in '([{':
            stack.push(char)
        elif char in ')]}':
            if stack.is_empty() or stack.pop() != matching[char]:
                return False

    return stack.is_empty()


def demonstrate_stack():
    print("=== Stack (LIFO) ===\n")

    s = Stack()
    s.push(1)
    s.push(2)
    s.push(3)
    print(s)
    print("Peek       :", s.peek())     # 3
    print("Pop        :", s.pop())      # 3
    print("Pop        :", s.pop())      # 2
    print("Size       :", s.size())     # 1
    print("Is empty?  :", s.is_empty()) # False
    s.pop()
    print("Is empty?  :", s.is_empty()) # True

    print("\n--- Balanced parentheses ---")
    tests = ["([]{})", "([)]", "((())", "{[()]}"]
    for expr in tests:
        print(f"  {expr!r:10s} → {is_balanced(expr)}")


if __name__ == "__main__":
    demonstrate_stack()

File: java/Stack.java
/**
 * Stack (Last-In, First-Out – LIFO) in Java
 * ==========================================
 * A stack lets you add and remove items only from the top.
 * Implemented here with a singly linked list (no array resizing needed).
 *
 * Real-world uses:
 *   - Undo / redo in editors
 *   - Browser back navigation
 *   - Function call stack
 *   - Balanced parentheses checking
 *
 * Complexities (all O(1)):
 *   push, pop, peek, isEmpty, size
 */

public class Stack {

    // ── Internal Node ────────────────────────────────────────────────────────
    private static class Node {
        int data;
        Node next;
        Node(int data) { this.data = data; }
    }

    private Node top;
    private int size;

    /** Add item to the top – O(1). */
    public void push(int item) {
        Node newNode = new Node(item);
        newNode.next = top;
        top = newNode;
        size++;
    }

    /** Remove and return the top item – O(1). */
    public int pop() {
        if (isEmpty()) throw new java.util.EmptyStackException();
        int value = top.data;
        top = top.next;
        size--;
        return value;
    }

    /** Return the top item without removing it – O(1). */
    public int peek() {
        if (isEmpty()) throw new java.util.EmptyStackException();
        return top.data;
    }

    /** Return true if the stack is empty – O(1). */
    public boolean isEmpty() { return top == null; }

    /** Return the number of items – O(1). */
    public int size() { return size; }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder("Stack (top→bottom): [");
        Node current = top;
        while (current != null) {
            sb.append(current.data);
            if (current.next != null) sb.append(", ");
            current = current.next;
        }
        return sb.append("]").toString();
    }

    // ── Practical example: balanced parentheses ──────────────────────────────
    public static boolean isBalanced(String expr) {
        Stack stack = new Stack();
        for (char ch : expr.toCharArray()) {
            if (ch == '(' || ch == '[' || ch == '{') {
                stack.push(ch);
            } else if (ch == ')' || ch == ']' || ch == '}') {
                if (stack.isEmpty()) return false;
                char open = (char) stack.pop();
                if ((ch == ')' && open != '(') ||
                    (ch == ']' && open != '[') ||
                    (ch == '}' && open != '{')) return false;
            }
        }
        return stack.isEmpty();
    }

    // ── Demo ─────────────────────────────────────────────────────────────────
    public static void main(String[] args) {
        System.out.println("=== Stack (LIFO) ===\n");

        Stack s = new Stack();
        s.push(1); s.push(2); s.push(3);
        System.out.println(s);
        System.out.println("Peek     : " + s.peek());    // 3
        System.out.println("Pop      : " + s.pop());     // 3
        System.out.println("Pop      : " + s.pop());     // 2
        System.out.println("Size     : " + s.size());    // 1
        System.out.println("IsEmpty  : " + s.isEmpty()); // false
        s.pop();
        System.out.println("IsEmpty  : " + s.isEmpty()); // true

        System.out.println("\n--- Balanced parentheses ---");
        String[] tests = {"([]{})", "([)]", "((())", "{[()]}"};
        for (String t : tests) {
            System.out.printf("  %-10s → %b%n", t, isBalanced(t));
        }
    }
}
